Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift

```diff
@@ -0,0 +1,133 @@
+  ///   case admin
+  ///   case guest
+  ///   case member
+  /// }
+  ///
+  /// var input = "admin,123"[...]
+  /// try Role.parser().parse(&input)  // Role.admin
+  /// input                            // ",123"
+  /// ```
+  ///
+  /// - Parameter inputType: The `Substring` type. This parameter is included to mirror the
+  ///   interface that parses any collection of UTF-8 code units.
+  /// - Returns: A parser that consumes a case-iterable, raw representable value from the beginning
+  ///   of a substring.
+  @inlinable
+  public static func parser(
+    of inputType: Substring.Type = Substring.self
+  ) -> Parsers.CaseIterableRawRepresentableParser<Substring, Self, String> {
+    .init(toPrefix: { $0 }, areEquivalent: ==)
+  }
+
+  /// A parser that consumes a case-iterable, raw representable value from the beginning of a
+  /// collection of a substring's UTF-8 view.
+  ///
+  /// - Parameter inputType: The `Substring.UTF8View` type. This parameter is included to mirror the
+  ///   interface that parses any collection of UTF-8 code units.
+  /// - Returns: A parser that consumes a case-iterable, raw representable value from the beginning
+  ///   of a substring's UTF-8 view.
+  @inlinable
+  public static func parser(
+    of inputType: Substring.UTF8View.Type = Substring.UTF8View.self
+  ) -> Parsers.CaseIterableRawRepresentableParser<Substring.UTF8View, Self, String.UTF8View> {
+    .init(toPrefix: { $0.utf8 }, areEquivalent: ==)
+  }
+
+  /// A parser that consumes a case-iterable, raw representable value from the beginning of a
+  /// collection of UTF-8 code units.
+  ///
+  /// - Parameter inputType: The collection type of UTF-8 code units to parse.
+  /// - Returns: A parser that consumes a case-iterable, raw representable value from the beginning
+  ///   of a collection of UTF-8 code units.
+  @inlinable
+  public static func parser<Input>(
+    of inputType: Input.Type = Input.self
+  ) -> Parsers.CaseIterableRawRepresentableParser<Input, Self, String.UTF8View>
+  where
+    Input.SubSequence == Input,
+    Input.Element == UTF8.CodeUnit
+  {
+    .init(toPrefix: { $0.utf8 }, areEquivalent: ==)
+  }
+}
+
+extension Parsers {
+  public struct CaseIterableRawRepresentableParser<
+    Input: Collection, Output: CaseIterable & RawRepresentable, Prefix: Collection
+  >: Parser
+  where
+    Input.SubSequence == Input,
+    Output.RawValue: Comparable,
+    Prefix.Element == Input.Element
+  {
+    @usableFromInline
+    let cases: [Output]
+
+    @usableFromInline
+    let toPrefix: (Output.RawValue) -> Prefix
+
+    @usableFromInline
+    let areEquivalent: (Input.Element, Input.Element) -> Bool
+
+    @usableFromInline
+    init(
+      toPrefix: @escaping (Output.RawValue) -> Prefix,
+      areEquivalent: @escaping (Input.Element, Input.Element) -> Bool
+    ) {
+      self.toPrefix = toPrefix
+      self.areEquivalent = areEquivalent
+      self.cases = Output.allCases.sorted(by: { $0.rawValue > $1.rawValue })
```

Comment:
Wouldn't it be sufficient to sort by length instead? This would allow dropping the `Comparable` requirement, isn't it?

Reply:
@tgrapperon as in something like this?

```swift
self.cases = Output.allCases.sorted(by: {
  toPrefix($0.rawValue).endIndex > toPrefix($1.rawValue).endIndex
})
```

Reply:
I guess it doesn't really matter if they have the same length. The main issue would be if a shorter value is ordered before a longer one. This probably means that comparing `.endIndex` would probably work for `Double` but comparing `.rawValue` might not

Reply:
@tgrapperon actually the current implementation would fail for negative `Int`s too. Maybe using `endIndex` is actually the way to go. You can change the current test to negative values to test it out. 

```swift
enum Person: Int, CaseIterable {
  case blob = -4
  case blobJr = -42
}

var input = "-4,-42"[...].utf8
XCTAssertEqual(try peopleParser.parse(&input), [.blob, .blobJr]) // XCTAssertEqual failed
```

Reply:
@iampatbrown With the current state of the PR, this is not really a concern. Sorting is useless but mostly harmless for numeric `RawValue`s, and the other types are collections where sorting by `count` would be equivalent to bin the `lexical` sort.
So it's working indeed, but it performs a little more work than needed. The `Comparable` requirement is fortuitously fulfilled by all supported `RawValue` types. Being `LosslessStringConvertible` automatically provides a canonical way to sort values by projecting them to `.description`, so it can maybe be possible to drop the `Comparable` requirement for a simple `LosslessStringConvertible`, and perform the sort via `.description` only when needed if we manage to exclude numbers (maybe by checking if they're `IntegerLiteralType` or `FloatLiteralType`, or using explicit types directly).

Reply:
> actually the current implementation would fail for negative Ints too.
Good catch!
As I'm getting it, it would work using `.description`, no? I didn't have the time to check out the branch yet.

Reply:
@iampatbrown Ok, I've just checked, and reverse sorting lexically by `.description` seems to do the trick too, but the `toPrefix` seems more efficient/elegant. It doesn't seem to cause issues with same-length strings because they're always different numbers(e.g., "-4" vs "44").

Reply:
Well the sort happens inside the init of the parser where `RawValue` has no constraints apart from being comparable. The `LosslessStringConvertible` was only for a static initilizer. I think you're right about using length though. The sort is to ensure the parser tries to match longer cases first. For `RawValue == String` using `Comparable` works fine.

Reply:
@iampatbrown That's a chance you did spot the negative `Int` case. This is a very common situation!